Add billing documents to Cosmos DB

mbarnes opened the pull request:
Jira: unknown

### What

I don't know if this has been carded up in Jira yet, but there's a [long-running Slack thread about billing](https://redhat-external.slack.com/archives/C075PHEFZKQ/p1739288836782679) where we pretty much worked out all the details with the Microsoft team implementing the billing service.

The RP side is fairly simple.  The backend component creates a new billing document upon successful completion of a cluster creation.  When the cluster is deleted, the backend updates the same billing document with a deletion timestamp.

### Why

See the [ARO-HCP OpenShift License Fee Billing Design Doc ](https://microsoft-my.sharepoint.com/:w:/r/personal/gniranjan_microsoft_com/_layouts/15/Doc.aspx?sourcedoc=%7B00BBF806-45C7-43D5-8268-24A9D278DB22%7D&file=ARO-HCP%20OpenShift%20License%20Fee%20Billing%20Design%20Doc.docx&nav=eyJjIjo3NDkzMDQyODB9&action=default&mobileredirect=true&share=IQEG-LsAx0XVQ4JoJKnSeNsiAXT6HBRQu9xQ_dRwX3lqw2A) for an overview, the shared Cosmos DB document structure, and RP requirements.

There's a couple deviations in the final Cosmos DB document structure:
- The addition of a ~~`partitionKey`~~ (now `subscriptionId`) JSON field, the value of which is the Azure subscription ID.
- The time fields are RFC 3339 timestamps rather than integer Unix times in seconds.
- We omit timestamps added by the billing service from our document definition, since the RP's interaction with the Billing container is write-only.

stevekuznetsov commented on internal/database/database.go:
```diff
@@ -178,6 +189,7 @@ func NewDBClient(ctx context.Context, database *azcosmos.DatabaseClient) (DBClie
 	// NewContainer only fails if the container ID argument is
 	// empty, so we can safely disregard the error return value.
 	resources, _ := database.NewContainer(resourcesContainer)
+	billing, _ := database.NewContainer(billingContainer)
```
This method already returns an error, being crafty and assuming the ways in which this fails (in the current implementation, not necessarily in the future) IMO just seems like a bad idea. Very simple to check the error and return.

mbarnes replied:
That's fair.  Added a new commit to fix this.

stevekuznetsov commented on internal/database/database.go:
```diff
@@ -204,6 +217,85 @@ func (d *cosmosDBClient) GetLockClient() *LockClient {
+func (d *cosmosDBClient) getBillingID(ctx context.Context, resourceID *azcorearm.ResourceID) (string, error) {
+	pk := NewPartitionKey(resourceID.SubscriptionID)
+
+	// Resource ID alone does not uniquely identify a billing document, but
```
What is the primary key?

mbarnes replied:
Primary key is the `ID` field.

mbarnes replied:
Correct.  Maybe it won't be an issue but I didn't want to assume the billing service will expunge a billing document for a deleted cluster before the RP might have to create a new one with the same resource ID.

mbarnes replied:
I _think_ I answered your question but feel free to unresolve if additional discussion is needed.

stevekuznetsov commented on internal/database/database.go:
```diff
@@ -204,6 +217,85 @@ func (d *cosmosDBClient) GetLockClient() *LockClient {
+			return "", fmt.Errorf("failed to advance page while querying Billing container for '%s': %w", resourceID, err)
+		}
+
+		for _, item := range queryResponse.Items {
```
Shouldn't a length >1 be an error here?

mbarnes replied:
Perhaps, but then that means never recording a deletion timestamp and we would potentially continue billing a deleted cluster.

stevekuznetsov replied:
I'm assuming that resource ID is not the primary key since it contains the semantic name of the cluster, and someone deleting it then quickly re-creating it would lead to more than one entry in the database for this resource ID, just with different creation timestamps. Since you also `AND NOT IS_DEFINED(c.deletionTime)`, the only case where we end up with more than one item in this query is, what, a race condition where the user is deleting the previous cluster and creating the current one immediately, and we have not yet updated the previous deletion time, but have already written the new one to the DB? 

Marking the new cluster as deleted sounds like the wrong thing to do there. 

Grabbing the first (sorted how?) document and returning that billing ID also seems wrong, seems like we need to carefully reason about what happened to get us into this state. Fail loudly when N > 1 and alert something, if putting in some 404 transient is not sufficient to eventually get to the right outcome. Silently doing "something" seems like the least best option.

mbarnes replied:
> Shouldn't a length >1 be an error here?
Addressed this much in a new commit.

mbarnes commented on backend/operations_scanner.go:
```diff
@@ -797,6 +822,52 @@ func (s *OperationsScanner) postAsyncNotification(ctx context.Context, op operat
+	doc.ManagedResourceGroup = fmt.Sprintf(
+		"/%s/%s/%s/%s",
+		azcorearm.SubscriptionResourceType.Type,
+		op.doc.ExternalID.SubscriptionID,
```
Subscription ID can be extracted from the `ResourceID` field.  Ok by me if we just want to record the MRG name, but I'd prefer to change the field name then to `ManagedResourceGroupName` (like CS has) for clarity.

mbarnes replied:
> The reason is, on billing service, we use the ARM resource ID of ManagedResourceGroup as a key to store and process events for any given hcp cluster.

Is that going to cause problems if a cluster is deleted and recreated with a managed resource group of the same name?  Or will billing always process the cluster deletion before the new one appears?  This is similar to a concern I had for _cluster_ resource IDs.

That aside, there's a `partitionKey` field whose value _happens to be_ the Azure subscription ID.  How about I rename it to `subscriptionId` so it's explicit and I'll reconfigure the container to use that field as the partition key.  So same partition values, but clearer names.

gouthamMN replied:
> Is that going to cause problems if a cluster is deleted and recreated with a managed resource group of the same name? Or will billing always process the cluster deletion before the new one appears? This is similar to a concern I had for cluster resource IDs.

I can't think of anything that cause issue from the billing perspective as the new cluster is under the same CX (subId) bcz of the following reasons.
a. **Process events:**
`HOBO/License fee billing:` 
1. Reads billing docs from CosmosDB.
2. Simply reads the VM compute events under the `managedResourceGroup`.

`Hosting fee`:
1. Reads billing docs from CosmosDB.
2. For each billing doc, creates a hosting usage event for each hour since `lastBillingTime` to a point recently received a VM usage event `BatchProcessingTime` for both old and new billing docs.

b. **Update Billing Docs:**
4. Update the `lastBillingTime` for each billing doc.
5. Deletes the billing doc if deletion time is set and `deletionTime` is before `lastBillingTime`. This won't impact the new billing doc as the deletion time is not set in that.

@stevekuznetsov - any thoughts? Also, I was thinking if it's safe to maintain a separate field to track the last billing time of Hosting fee (decouple it from `BatchProcessingTime` of VM compute events). If we create dependency on `BatchProcessingTime` this will impact hosting fee in case of Pav2 fails send events for whatever reason.

@mbarnes - Although I don't see any issues today, I suggest maintaining a unique name for MRG under same SubId if it's possible.

gouthamMN replied:
> That aside, there's a partitionKey field whose value happens to be the Azure subscription ID. How about I rename it to subscriptionId so it's explicit and I'll reconfigure the container to use that field as the partition key. So same partition values, but clearer names.

I'm good with either way for this.

mbarnes replied:
> @mbarnes - Although I don't see any issues today, I suggest maintaining a unique name for MRG under same SubId if it's possible.

If I'm understanding you correctly I don't think we can guarantee this because customers can pick their own managed resource group name.  I think you'd have to factor in the creation timestamp to get a truly unique identifier.

mbarnes replied:
> I'm good with either way for this.

Ok I'll leave the fields alone for now.  But if you find yourself wanting to grab the subscription ID value from  `partitionKey` then I'd rather change the name.

gouthamMN replied:
@mbarnes - can you add `lastBillingTimeOfMngtEvent` & rename `lastBillingTime` to `lastBillingTimeOfVMEvent`. The former will be used for processing Hosting/Management fee and later will be used for processing license fee.

gouthamMN replied:
@mbarnes - Also, looks like we will be using `subID`, feel free to rename `partitionKey`.

mbarnes replied:
> @mbarnes - can you add lastBillingTimeOfMngtEvent & rename lastBillingTime to lastBillingTimeOfVMEvent. The former will be used for processing Hosting/Management fee and later will be used for processing license fee.

Actually... as far as the RP is concerned, having `omitempty` on `lastBillingTime` is the same as just omitting it from the definition. The RP doesn't write that field nor does it ever read the full document back in (it patches in `deletionTime`).  So I think I'll just remove `lastBillingTime` and you can inject whatever additional timestamps you need on your end.  These _are_ just JSON objects, after all.  There's no schema.

gouthamMN replied:
That's perfect!!!

mbarnes replied:
Ok, database changes are made.  I deleted `lastBillingTime` from the RP's definition and renamed `partitionKey` to `subscriptionId`.  So the partition key path is now `/subscriptionId`.  If you have a test environment up already you'll need to delete and recreate the Billing container for the partition key change to take effect.

gouthamMN reviewed (commented):
lgtm

stevekuznetsov commented on backend/operations_scanner.go:
```diff
@@ -481,6 +484,17 @@ func (s *OperationsScanner) pollClusterOperation(ctx context.Context, op operati
+			err = s.patchBillingDocument(ctx, op)
+			if err != nil {
+				s.recordOperationError(ctx, pollClusterOperationLabel, err)
+				op.logger.Error(fmt.Sprintf("Failed to handle a completed deletion: %v", err))
```
Let's distinguish these so the poor sod trying to make sense of what failed, knows.

```suggestion
				op.logger.Error(fmt.Sprintf("Failed to update billing document after a completed deletion: %v", err))
```

stevekuznetsov commented on backend/operations_scanner.go:
```diff
@@ -501,6 +515,19 @@ func (s *OperationsScanner) pollClusterOperation(ctx context.Context, op operati
+	if op.doc.Request == database.OperationRequestCreate && opStatus == arm.ProvisioningStateSucceeded {
+		err = s.createBillingDocument(ctx, op)
+		if err != nil {
+			s.recordOperationError(ctx, pollClusterOperationLabel, err)
```
Why `pollClusterOperationLabel`?

mbarnes commented on backend/operations_scanner.go:
```diff
@@ -481,6 +484,17 @@ func (s *OperationsScanner) pollClusterOperation(ctx context.Context, op operati
+			// document still having a non-terminal status.
+			err = s.patchBillingDocument(ctx, op)
+			if err != nil {
+				s.recordOperationError(ctx, pollClusterOperationLabel, err)
```
Metrics are @simonpasquier's domain, but basically any errors that occur within `pollClusterOperation` use this label.  Unclear if we would want additional metric labels for billing operations.  I'd defer to Simon.

mbarnes commented on backend/operations_scanner.go:
```diff
@@ -820,6 +847,52 @@ func (s *OperationsScanner) postAsyncNotification(ctx context.Context, op operat
+	if err == nil {
+		op.logger.Info("Updated billing for cluster deletion")
+	} else if database.IsResponseError(err, http.StatusNotFound) {
+		// Log the error but proceed with normal processing.
```
Probably a metric, I don't think we would alert on log messages.  At least for this initial cut I had existing development environments in mind where clusters have already been created without billing records.

mbarnes commented on backend/operations_scanner.go:
```diff
@@ -501,6 +515,19 @@ func (s *OperationsScanner) pollClusterOperation(ctx context.Context, op operati
+		err = s.createBillingDocument(ctx, op)
+		if err != nil {
+			s.recordOperationError(ctx, pollClusterOperationLabel, err)
+			op.logger.Error(fmt.Sprintf("Failed to handle a completed creation: %v", err))
```
Similar to above comment, the wrapped error has this information.
The full log message would be something like:
```
Failed to handle a completed creation: failed to create Billing container item for '<cluster-resource-id>': <cosmos-db-error>
```

mbarnes commented on backend/operations_scanner.go:
```diff
@@ -820,6 +847,52 @@ func (s *OperationsScanner) postAsyncNotification(ctx context.Context, op operat
+
+// patchBillingDocument patches a Cosmos DB document in the Billing
+// container to add a deletion timestamp.
+func (s *OperationsScanner) patchBillingDocument(ctx context.Context, op operation) error {
```
Updated the last commit with the name change.

mbarnes commented on backend/operations_scanner.go:
```diff
@@ -820,6 +847,52 @@ func (s *OperationsScanner) postAsyncNotification(ctx context.Context, op operat
+func (s *OperationsScanner) patchBillingDocument(ctx context.Context, op operation) error {
+	var patchOperations database.BillingDocumentPatchOperations
+
+	patchOperations.SetDeletionTime(time.Now().UTC())
```
Appended a new commit for this so you can verify.

mbarnes commented:
Addressed all the review comments.  I'm inclined to defer adding new metrics for now since this is just a first cut at billing integration.  Let's make sure this works well with the billing service first, then we can think about alerts.
